### DISCUSSION TITLE
How do I run the most recent Linux x64 beta release on an external server, such as one hosted by BisectHosting?

### DISCUSSION THREAD

**Discussion by PixelGMS** [Pryaxis/TShock#2738] [Q&A]
I've tried several different things, but each time I get different errors.
I've tried deleting all the files in the file manager and replacing them with the contents of the .tar, I've tried doing that except leaving the executable, I've tried only deleting the files that had versions in the .tar, and I tried more things three weeks ago before giving up and deciding to wait to see if the 1.4.4 TShock prerelease would be released anytime soon.
I have noticed that the external server has a TerrariaServer.pdb and TerrariaServer.exe file whereas the beta has neither of those but has a TShock.server which the BisectHosting server doesn't.
Before you ask, I did ask BisectHosting's customer server for help before coming here, they didn't have any idea of what to do.
Thank you in advance for any help given.

**Comment by Arthri** [comment]
> I've tried several different things, but each time I get different errors.

can you name a few of the errors?

> I have noticed that the external server has a TerrariaServer.pdb and TerrariaServer.exe file 

are you able to replace these?

TShock 5's server structure is different from TShock 4. To move from TShock 4 to 5 on an external host you need to be able to
- Replace all the files in the server directory including TerrariaServer.exe
- Install software(.NET 6 x64 Console Runtime) if necessary

**Reply by PixelGMS** [reply]
I can edit, delete, download, and upload files freely but the startup command is locked as "mono TerrariaServer.exe -config config.txt" and I can't change it, though if it's necessary I could send a support ticket to BisectHosting about it. Or, I can change it, but only to presets for vanilla Terraria, TShock 1.4.3.6, and various versions of TModLoader. 
Startup command for vanilla Terraria: bash TerrariaServer -config config.txt
Startup command for TModLoader V1.3: bash tModLoaderServer -config config.txt
Startup command for TModLoader V1.4: ./LaunchUtils/ScriptCaller.sh -server -config config.txt -lobby friends
Startup command for TModLoader 64Bit: mono tModLoader64BitServer.exe -config config.txt
I'm pretty sure I won't be able to install software, but if it's necessary I could send a support ticket about it.

Console when I delete all the files and then replace them by the beta.
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Cannot open assembly 'TerrariaServer.exe': No such file or directory.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Cannot open assembly 'TerrariaServer.exe': No such file or directory.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

Console when I delete all the files except for TerrariaServer.exe and then unzip the tar file
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Unhandled Exception:
System.TypeLoadException: Could not load type of field 'TerrariaApi.Server.ServerApi:game' (2) due to: Could not load file or assembly 'OTAPI, Version=1.4.3.6, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeLoadException: Could not load type of field 'TerrariaApi.Server.ServerApi:game' (2) due to: Could not load file or assembly 'OTAPI, Version=1.4.3.6, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 1
[Panel]: Out of memory: false
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Unhandled Exception:
System.TypeLoadException: Could not load type of field 'TerrariaApi.Server.ServerApi:game' (2) due to: Could not load file or assembly 'OTAPI, Version=1.4.3.6, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeLoadException: Could not load type of field 'TerrariaApi.Server.ServerApi:game' (2) due to: Could not load file or assembly 'OTAPI, Version=1.4.3.6, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 1
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

Console when I only delete files that also exist in the beta's tar file
TerrariaAPI Version: 2.1.0.0 (Protocol v1.4.3.6 (248), OTAPI 1.4.3.6)
[Server API] Error Startup aborted due to an exception in the Server API initialization:
System.InvalidOperationException: Failed to load assembly "TShockAPI.dll". ---> System.InvalidOperationException: Could not create an instance of plugin class "TShockAPI.TShock". ---> System.TypeLoadException: Invalid type Rests.SecureRest for instance field TShockAPI.TShock:RestApi
  at (wrapper managed-to-native) System.RuntimeType.GetConstructors_native(System.RuntimeType,System.Reflection.BindingFlags)
  at System.RuntimeType.GetConstructors_internal (System.Reflection.BindingFlags bindingAttr, System.RuntimeType reflectedType) [0x00008] in <de0ee687c1ad4ea99fca2db5d4839ca1>:0 
  at System.RuntimeType.GetConstructorCandidates (System.String name, System.Reflection.BindingFlags bindingAttr, System.Reflection.CallingConventions callConv, System.Type[] types, System.Boolean allowPrefixLookup) [0x00034] in <de0ee687c1ad4ea99fca2db5d4839ca1>:0 
  at System.RuntimeType.GetConstructors (System.Reflection.BindingFlags bindingAttr) [0x00000] in <de0ee687c1ad4ea99fca2db5d4839ca1>:0 
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x0022b] in <de0ee687c1ad4ea99fca2db5d4839ca1>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x00095] in <de0ee687c1ad4ea99fca2db5d4839ca1>:0 
  at System.Activator.CreateInstance (System.Type type, System.Object[] args) [0x00000] in <de0ee687c1ad4ea99fca2db5d4839ca1>:0 
  at TerrariaApi.Server.ServerApi.LoadPlugins () [0x00232] in <c8b4d495d61446e9a784fcc2bffb6307>:0 
   --- End of inner exception stack trace ---
  at TerrariaApi.Server.ServerApi.LoadPlugins () [0x0027b] in <c8b4d495d61446e9a784fcc2bffb6307>:0 
   --- End of inner exception stack trace ---
  at TerrariaApi.Server.ServerApi.LoadPlugins () [0x002ba] in <c8b4d495d61446e9a784fcc2bffb6307>:0 
  at TerrariaApi.Server.ServerApi.Initialize (System.String[] commandLineArgs, Terraria.Main game) [0x0011a] in <c8b4d495d61446e9a784fcc2bffb6307>:0 
  at TerrariaApi.Server.HookManager.InitialiseAPI () [0x0004e] in <c8b4d495d61446e9a784fcc2bffb6307>:0 
I then pressed stop and the server suddenly began to start but it was v1.4.3.6 not v1.4.4.x

**Reply by Arthri** [reply]
I suppose you can try picking vanilla terraria then renaming TShock.Server.exe to TerrariaServer.exe

**Reply by PixelGMS** [reply]
I tried renaming TShock.Server (It wasn't TShock.Server.exe, just TShock.Server) to TerrariaServer.exe. I got the following error message.
To clarify, I turned the server into a vanilla server and then uploaded the contents of the tar and then renamed the file.
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ bash TerrariaServer -config config.txt
Cannot open assembly 'TerrariaServer.exe': File does not contain a valid CIL image.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ bash TerrariaServer -config config.txt
Cannot open assembly 'TerrariaServer.exe': File does not contain a valid CIL image.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

**Reply by Arthri** [reply]
Oof. This likely means .NET 6 Runtime isn't installed. You can either request for it to get installed or clone TShock and recompile TShock as a self-contained app

**Reply by PixelGMS** [reply]
How do I clone TShock and recompile it as a self-contained app?

**Reply by Arthri** [reply]
You need .NET 6 x64 SDK. To clone TShock, you can either use git/gh CLI or download the [zipped version](https://github.com/Pryaxis/TShock/archive/refs/heads/general-devel.zip). After you've cloned it, open a terminal inside it and run `cd TShockLauncher && dotnet publish -r RUNTIMEIDENTIFIER --sc true -f net6.0 -c Release`. Replace `RUNTIMEIDENTIFIER` with the appropriate one for your OS. A list of them is available at https://learn.microsoft.com/en-us/dotnet/core/rid-catalog. The output should then be available under `TShockLauncher/bin/Release/net6.0/win-x64/publish/`, you can zip this version and send it back onto the server. Do note it will be noticeably larger since it contains .NET 6 and TShock, rather than just TShock

Alternatively, you can request your host to install .NET 6 x64 Console Runtime

**Reply by PixelGMS** [reply]
To clarify, do I want to replace RUNTIMEIDENTIFIER with the appropriate one for my desktop's OS or for BisectHosting's OS? Logically I'm assuming BisectHosting's OS but the way you word it makes me unsure.

**Reply by Arthri** [reply]
Yup. Use BisectHosting's OS. If you want to run it in on your own machine, replace it with your OS'

**Reply by PixelGMS** [reply]
So, I need to install .NET 6 x64 SDK, right? So I install my desktop's version (Windows x64) of it and then I'll be able to run "cd TShockLauncher && dotnet publish -r linux-x64 --sc true -f net6.0 -c Release", right?

**Reply by Arthri** [reply]
yup

**Reply by PixelGMS** [reply]
I got a rather long error
[TShock Error.txt](https://github.com/Pryaxis/TShock/files/9834895/TShock.Error.txt)

**Reply by Arthri** [reply]
Oh. right. You also need to get [TerrariaServerAPI](https://github.com/Pryaxis/TSAPI/archive/refs/heads/general-devel.zip) as well and paste it into a "TerrariaServerAPI" folder under TShock. If you're using git, just open a terminal in the TShock folder and do `git submodule update --init --recursive`. Btw, if you want shorter errors do add `-p:NoWarn:CS1591` to the end of the `dotnet publish` command. This silences some irrelevant warnings

**Reply by PixelGMS** [reply]
If I'm not using git do I unzip it into the TerrariServerAPI or do I leave it in the zip?

**Reply by Arthri** [reply]
Unzip it

**Reply by PixelGMS** [reply]
And do I place that single folder into TerrariaServerAPI or do I put the contents of the folder into TerrariaServerAPI?

**Reply by Arthri** [reply]
put its contents into TerrariaServerAPI. the folder should look like this https://github.com/Pryaxis/TSAPI/

**Reply by PixelGMS** [reply]
Okay, so, I zipped everything in "TShockLauncher/bin/Release/net6.0/win-x64/publish/" and uploaded it to the server and tried deleting everything and uploading the contents of the zips for both when it had the startup commands for vanilla and Tshock, and did it again and renamed the TShock.Server to TerrariaServer and TerrariaServer.exe respectively and each time it crashed.

**Reply by Arthri** [reply]
What's the errors? Btw are you using any plugins?

**Reply by PixelGMS** [reply]
Sorry, didn't get a notification that you responded.
(The file I'm referring to being renamed is TShock.Server)
Vanilla, renamed file
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
chmod: cannot access 'TerrariaServer.bin.x86_64': No such file or directory
:/home/container$ bash TerrariaServer -config config.txt
TerrariaServer: TerrariaServer: cannot execute binary file
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 126
[Panel]: Out of memory: false
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
chmod: cannot access 'TerrariaServer.bin.x86_64': No such file or directory
:/home/container$ bash TerrariaServer -config config.txt
TerrariaServer: TerrariaServer: cannot execute binary file
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 126
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

Vanilla, file unrenamed
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
chmod: cannot access 'TerrariaServer': No such file or directory
chmod: cannot access 'TerrariaServer.bin.x86_64': No such file or directory
:/home/container$ bash TerrariaServer -config config.txt
bash: TerrariaServer: No such file or directory
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 127
[Panel]: Out of memory: false
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
chmod: cannot access 'TerrariaServer': No such file or directory
chmod: cannot access 'TerrariaServer.bin.x86_64': No such file or directory
:/home/container$ bash TerrariaServer -config config.txt
bash: TerrariaServer: No such file or directory
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 127
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

TShock renamed
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Cannot open assembly 'TerrariaServer.exe': File does not contain a valid CIL image.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

TShock Unrenamed
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Cannot open assembly 'TerrariaServer.exe': No such file or directory.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Cannot open assembly 'TerrariaServer.exe': No such file or directory.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

**Reply by PixelGMS** [reply]
And no, I'm not using any plugins.

**Reply by PixelGMS** [reply]
Though there are stuff in the plugins server.
TShockAPI.deps.json
TShockAPI.dll
TShockAPI.pdb
TShockAPI.xml

**Reply by Arthri** [reply]
Do you have a bin folder beside the plugins folder? try pasting the contents of the bin folder into the root directory

**Reply by PixelGMS** [reply]
There's a publish folder with a bin folder, a copy of the ServerPlugins folder with all the same files (or, at least, they have the same names), a TShock.server file, a TShock.server.deps.json, TShock.server.dll, and TShock.server.runtimeconfig.json file in it.
You want me to move the contents of the bin folder, which is in the publish folder, into the root directory?

**Reply by Arthri** [reply]
No. you're supposed to use the publish folder. then move publish/bin to just publish and push publish to the server

**Reply by PixelGMS** [reply]
Got this after I did that.
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Cannot open assembly 'TerrariaServer.exe': File does not contain a valid CIL image.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
:/home/container$ mono TerrariaServer.exe -config config.txt
Cannot open assembly 'TerrariaServer.exe': File does not contain a valid CIL image.
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 2
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.
Note: At first I accidentally pushed the contents of publish into the root directory before realizing my mistake and pushing the folder. Should I delete everything in root and start from unzipping the contents of the linux-x64 in root again?

**Reply by Arthri** [reply]
btw it needs to run through bash, or anything except mono really. TShock 5 uses .NET 6, which is different and incompatible with mono

**Reply by PixelGMS** [reply]
Is there any way to get around that? Or should I just send a ticket to BisectHosting to change the startup command, and, if so, what to?

**Reply by Arthri** [reply]
If you set the preset to vanilla it runs through bash

**Reply by PixelGMS** [reply]
Okay, and then I want to rename TShock.Server to TerrariaServer?

**Reply by Arthri** [reply]
yup

**Reply by PixelGMS** [reply]
[Panel]: Checking server disk space usage, this could take a few seconds...
[Panel]: Updating process configuration files...
[Panel]: Ensuring file permissions are set correctly, this could take a few seconds...
[Panel]: Server marked as starting...
[Panel]: Pulling Docker container image, this could take a few minutes to complete...
[Panel]: Finished pulling Docker container image
chmod: cannot access 'TerrariaServer.bin.x86_64': No such file or directory
:/home/container$ bash TerrariaServer -config config.txt
TerrariaServer: TerrariaServer: cannot execute binary file
[Panel]: Server marked as offline...
[Panel]: ---------- Detected server process in a crashed state! ----------
[Panel]: Exit code: 126
[Panel]: Out of memory: false
[Panel]: Aborting automatic restart, last crash occurred less than 60 seconds ago.

**Reply by Arthri** [reply]
Ah I see. I have no idea. I suppose the other option is sending a request ticket for .NET 6 runtime

**Reply by PixelGMS** [reply]
If I get them to install .NET 6 runtime what do I need to do?

**Reply by Arthri** [reply]
If you get them to install .NET 6 you can use stock TShock 5(or your own whichever you prefer) through vanilla preset / bash
